CRAMcodecs.tex

```diff
@@ -1721,7 +1740,7 @@ \subsection{Statistical Modelling}
 frequency tables in the data stream covering all possible contexts.
 
 To do this we use a statistical model, containing an array of symbols $S$ and their frequencies $F$.
-The sum of these frequences must be less than $2^{16}-32$.
+The sum of these frequences must be less than $2^{16}-16$ so after adjusting the frequencies it never go above the maximum unsigned 16-bit integer.
```

Comment:
The sum of these frequences must be less than or equal to $2^{16}-16$ so after adjusting the frequencies it never go above the maximum unsigned 16-bit integer.

Reply:
We add 16 to the frequency, and $2^{16}-1$ is the maximum value stored in a uint16_t, not $2^{16}$.  Hence less than instead of less than or equal.

NB: it's incremented by 16 instead of 1 because we start with all symbols having a minimum probability of 1/N.  We want to distinguish those symbols seen and those not yet seen, so the jump from 1/N (1/256 for a full alphabet) to 17/272 and 33/288 is a more rapid learning of probabillities than simple +1.  The price though is more frequent renormalisation, but that's little more than a bunch of bit shifts.  It's better too for smaller N too, so limited alphabets benefit from less overhead for the as-yet unseen symbols.

Some range coders use an escape code and freq 0 for all unused symbols.  When we add a symbol that currently has freq 0, the escape code is emitted, then the new symbol itself, and then it's incremented as before.  Ie we start with 1/1 (P(escape)=1), then emit escape + 'x' and we have P(escape)=1/2 and P(x)=1/2.  Then another x => P(escape)=1/3, P(x)=2/3.  (Escape and literal symbols can be adjusted further based on the chance of new symbols vs how many emitted so far given a finite alphabet, but it's relatively minor).  This method is a better estimation of probabilities, but it's also more complex and I didn't deem the speed hit to be worth the small size difference, especially when we have a small alphabet.  It only really helps significantly on small blocks of data or those with non-stationary probabilities, where the size / speed of adaption matters the most.